### PR TITLE
[dynamo] Refactor DynamicShapeVariable to be a top level VariableTracker

### PR DESCRIPTION
**Pull request by voznesenskym**
cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @jansel

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -590,6 +607,33 @@ def python_type(self):
     def unpack_var_sequence(self, tx):
         super(DynamicShapeVariable, self).unpack_var_sequence(tx)
 
+    def as_python_constant(self):
+        return self.dyn_shape
```
Comment (voznesenskym) [inline torch/_dynamo/variables/tensor.py:611]: vaguely sus btw, looking at pulling this

Comment (voznesenskym) [reply torch/_dynamo/variables/tensor.py:611]: Yea pulling this gives us some real failures where we make consts erronously

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -219,7 +220,14 @@ def wrap_name(module_key):
 
             def wrap_name(module_key):
                 return NNModuleVariable(type(mod), module_key, **options)
-
+        elif isinstance(mod, torch.SymIntNode):
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:223]: huh, what, we're storing SymInts on a module? That's weird

File: torch/_dynamo/variables/builder.py [hunk 3]
```diff
@@ -490,6 +492,28 @@ def tensor_should_specialize(self):
             )
         )
 
+    def wrap_sym(self, value: Union[torch.SymIntNode, torch.SymFloatNode]):
+        if not is_constant_source(self.get_source()):
+            self.tx.output.graphargs.append(GraphArg(self.get_source(), value, False))
+        with torch._C.DisableTorchFunction():
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:498]: this shouldn't be necessary, I think

**Comment by ezyang** [comment]
idk how to review this lol

File: torch/_dynamo/output_graph.py (hunk 2 above)
Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:223]: No, this is not a module, at all. Let me change the type annotations, it's dead wrong.

**Review by jansel — APPROVED** [review]

**Comment by voznesenskym** [comment]
Integrated into symbolic-shapes, holding off land atm until that branch is all in

**Comment by ezyang** [comment]
No need to hold off, land this whenever possible

**Pull request closed without merging**